internal/statesync/syncer.go

```diff
@@ -264,8 +255,24 @@ func (s *syncer) Sync(ctx context.Context, snapshot *snapshot, chunks *chunkQueu
 		s.mtx.Unlock()
 	}()
 
+	hctx, hcancel := context.WithTimeout(ctx, 30*time.Second)
```

Comment:
Is there some reason to set an explicit timeout here rather than trusting the caller to plumb in a timeout? 

In my general experience, it can be difficult to debug failures when timeouts are added at arbitrary plumbing layers. The top-level caller probably has some idea how long they're willing to wait ("till the user is tired", "till the database needs a sync", or what have you), but if we try to guess inside that we're either guessing "too long" or risking cutting off a service request for a reason the caller wasn't interested in.

Reply:
Statesyncing is a cyclic process which involves continually trying to find snapshots and attempt to restore them if they fail. The top level user, depending on the estimated size of state may wait from minutes to hours for state sync to occur. This is an order of magnitude different to the time expected to produce data for verifying a snapshot. As we don't want to wait for hours on a specific snapshot, mid tier lever calls need to make decisions about when to give up on a snapshot and try with another one.